### ISSUE TITLE
Error using F.binary_cross_entropy in model.py with autocast

### ISSUE THREAD

**Issue by bsugerman** [VCasecnikovs/Yet-Another-YOLOv4-Pytorch#6]
This is a really great repo and I'm enjoying reading it. I've noticed it throws an error in lines 925:926 of model.py, since the target tensors are masks (either 0. or 1.). 

The complaint is that this function can't be autocast, and one should use F.binary_cross_entropy_with_logits instead. I'm curious if you have encountered this and whether it is worth updating your model.py to prevent this error?

Thanks!

**Comment by VCasecnikovs** [comment]
Hello, 
thank you for this issue.

As I understand you are referring to this lines of code:
```
loss_conf_obj = F.binary_cross_entropy(pred_conf[obj_mask], tconf[obj_mask])
loss_conf_noobj = F.binary_cross_entropy(pred_conf[noobj_mask], tconf[noobj_mask])
```
I haven't tried to use autocast on this model yet, because my colegue had disappointing results using half precision with OD tasks. As I understand, to make these changes we should replace pred_conf with prediction[..., 4] and also replace 1 with e and 0 with 1. I do not think that this operation would be so time-consuming to not do it.

I would be happy if you would do a pull request with an example of autocasting to this repository.

If you are referring to the iou_all_to_all function. This function calculates Intersection over union for two tensors of boxes. It is used to calculate iou for the nearest bboxes, which are its neighboring ground-truth objects, which are not its target, to make repulsion, and also for prediction bboxes to repulse prediction bbox from others with the different target. As it is described in https://arxiv.org/pdf/1711.07752.pdf.
I haven't tried to autocast this function and I do not understand how it can be replaced with binary_cross_entropy_with_logits.

Best regards,
Vadims.

**Comment by bsugerman** [comment]
Yes, those are the code lines.

> 
> As I understand you are referring to this lines of code:
> 
> ```
> loss_conf_obj = F.binary_cross_entropy(pred_conf[obj_mask], tconf[obj_mask])
> loss_conf_noobj = F.binary_cross_entropy(pred_conf[noobj_mask], tconf[noobj_mask])
> ```

pytorch autocasting works very well on codes I've used and developed, and yields a significant increase in speed and memory usage. I'm tinkering with using your code with autocasting and will try the different cross-entropy method and let you know if I find anything promising.

**Comment by VCasecnikovs** [comment]
Hope you will find something good. If you will have anything to pool request, I would be really happy.

**Comment by VCasecnikovs** [comment]
Maybe You had same as this #11 problem
